File: v1/testing/mWorld.cpp
// mWorld.cpp
#include "../header/World.h"
#include "../header/MakhlukLive.h"
#include "../header/LMakhluk.h"
#include "../header/KeypressHandler.h"
#include "../header/WorldBuilder.h"
#include "../header/Screen.h"
#include "../header/SnapshotCapturer.h"
#include "../header/ExceptionObject.h"
#include <iostream>
#include <string>
#include <algorithm>
#include <thread>
using namespace std;

int isValidID(string);
// Melakukan pengecekan string masukan pengguna, apakah ID yang dimasukkan benar (valid)

int main()
{
	/**************************************************************
	**
	**				Input dimensi ukuran dunia
	**
	**************************************************************/
	int _NBrs, _NKol;
	do
	{
		try
		{
			cout << "Masukkan dimensi ukuran ruang dunia." << endl;
			cout << "Jumlah baris (panjang) = "; cin >> _NBrs;
			cout << "Jumlah kolom (lebar)   = "; cin >> _NKol;
			if (_NBrs < 7 || _NKol < 7) throw ExceptionObject(0);
		}
		catch (ExceptionObject& E)
		{
			E.DisplayErrorMessage();
		}

	} while(_NBrs < 7 || _NKol < 7);
	World::getWorldInstance()->setNBrs(_NBrs+2); World::getWorldInstance()->setNKol(_NKol+2);

	cout << "\n===========================================================================\n" << endl;

	/**************************************************************
	**
	**				Input makhluk-makhluk yang ada
	**
	**************************************************************/
	string input;
	do
	{
		try
		{
			cout << "Masukkan sebuah string yang  berisi ID (UPPERCASE) makhluk-makhluk yang ingin dimunculkan pada dunia. " << endl;
			cout << "Pilihan :\n\tP = PolarBear, \n\tG = Tumbuhan, \n\tR = Rabbit, \n\tT = Turtle, \n\tS = Sheep, \n\tW = Wolf, \n\tU = Snake" << endl;
			cout << "String input = ";
			cin >> input;

			if(!isValidID(input)) throw ExceptionObject(1);
		}
		catch (ExceptionObject& E)
		{
			E.DisplayErrorMessage();
		}
	} while (!isValidID(input));
	WorldBuilder::getBuilderInstance()->setStrMakhluk(input);
	WorldBuilder::getBuilderInstance()->buildWorldObjects();

	cout << "\n===========================================================================\n" << endl;

	/**************************************************************
	**
	**					Nyalakan dunia
	**
	**************************************************************/

	int i =0;
    while (i==0) {
        MakhlukLive::getInstance()->MakhlukMove();
        Screen::getScreenInstance()->PrintWorldMap();
        MakhlukLive::getInstance()->MakhlukEat();
        std::chrono::milliseconds timespan(500);
        std::this_thread::sleep_for(timespan);
    }
	delete World::getWorldInstance();
	return 0;
}

int isValidID(string s)
{
	int Found = 0;
	for(int i=0; i<(int)s.size() && !Found; i++)
	{
		if(s[i] != 'P' && s[i] != 'G' && s[i] != 'R' && s[i] != 'T' && s[i] != 'S' && s[i] != 'W' && s[i] != 'U')
		{
			Found = 1;
		}
	}
	return !Found;
}

File: TubesMakhluk/test/MWorld.java
import java.util.Scanner;
import exception.ExceptionObject;
import inputoutput.Screen;
import makhluklive.MakhlukLive;
import world.World;
import world.WorldBuilder;

/** Kelas MWorld adalah kelas yang menampung main program.
 *  @author     Robert Sebastian Herlim / 13514061
 *  @version    1.0
 */
public class MWorld {
    /** fungsi validasi apakah id valid.
     * @param s string ID
     * @return boolean true jika ukuran valid
     */
    private static boolean isValidID(final String s) {
        boolean found = false;
        for (int i = 0; i < s.length() && !found; i++) {
            if (s.charAt(i) != 'P' && s.charAt(i) != 'G' && s.charAt(i) != 'R'
                    && s.charAt(i) != 'T' && s.charAt(i) != 'S'
                    && s.charAt(i) != 'W' && s.charAt(i) != 'U') {
                found = true;
             }
        }
        return !found;
    }
    /** fungsi validasi apakah ukuran valid.
     * @param strNBrs string jumlah baris
     * @param strNKol string jumlah kolom
     * @return boolean true jika ukuran valid
     */
     private static boolean isValidUkuran(final String strNBrs,
             final String strNKol) {
        boolean valid = true;
        for (int i = 0; i < strNBrs.length() && valid; i++) {
            if (strNBrs.charAt(i) < '0'
                    || strNBrs.charAt(i) > '9') {
                valid = false;
            }
        }
        for (int i = 0; i < strNKol.length() && valid; i++) {
            if (strNKol.charAt(i) < '0'
                    || strNKol.charAt(i) > '9') {
                valid = false;
            }
        }
        return valid;
     }
     /** Main program animal village.
      * @param args
      * @throws Exception
      */
     public static void main(final String args[]) throws Exception {
         System.out.println("***************************************"
                + "**********************************************");
         System.out.println("       **       ***    **  ***  ***    "
                + "     ***       **       **        **  ********");
         System.out.println("      ****      ****   **  ***  ****   "
                + "    ****      ****      **       **   **      ");
         System.out.println("     **  **     ** **  **  ***  ** **  "
                + "   ** **     **  **     **            **      ");
         System.out.println("    ********    **  ** **  ***  **  ** "
                + "  **  **    ********    **            ********");
         System.out.println("   **      **   **   ****  ***  **   **"
                + " **   **   **      **   **                  **");
         System.out.println("  **        **  **    ***  ***  **    *"
                + "**    **  **        **  ********      ********");
         System.out.println("                                      "
                + "                                               ");
         System.out.println("       **        **  ***  **        ** "
                + "            **       ********  ********       ");
         System.out.println("        **      **   ***  **        **"
                + "            ****      **        **             ");
         System.out.println("         **    **    ***  **        **"
                + "           **  **     **        **             ");
         System.out.println("          **  **     ***  **        **"
                + "          ********    **  ****  ********       ");
         System.out.println("           ****      ***  **        **"
                + "         **      **   **    **  **             ");
         System.out.println("            **       ***  ********  ***"
                + "*****  **        **  ********  ********       ");
         System.out.println("                                       "
                + "                                              ");
         System.out.println("             WOLF     RABBIT     TURTLE"
                + "     POLARBEAR     SNAKE     SHEEP            ");
         System.out.println("                                       "
                + "                                              ");
         System.out.println("                         BY EPIC: NGIONG,"
                + " MICKY, KEPI, ELVINA                        ");
         System.out.println("***************************************"
                + "**********************************************");
        /**************************************************************
        **
        **               Input dimensi ukuran dunia
        **
        **************************************************************/
        Scanner in = new Scanner(System.in);
        int nBrs = 0, nKol = 0;
        String strNBrs = "", strNKol = "";
        do {
           try {
               System.out.println("Masukkan dimensi ukuran"
                   + " ruang dunia.");
               System.out.print("Jumlah baris (panjang) = ");
               strNBrs = in.next();
               System.out.print("Jumlah kolom (lebar)   = ");
               strNKol = in.next();
               if (!isValidUkuran(strNBrs, strNKol)) {
                   throw new ExceptionObject(4);
               }
               nBrs = Integer.parseInt(strNBrs);
               nKol = Integer.parseInt(strNKol);
               if (nBrs < 7 || nKol < 7) {
                  throw new ExceptionObject(0);
               }
            } catch (ExceptionObject e) {
                e.displayErrorMessage();
            }
        } while (nBrs < 7 || nKol < 7
                || !isValidUkuran(strNBrs, strNKol));
        World.getWorldInstance().setnBrs(nBrs+2);
        World.getWorldInstance().setnKol(nKol+2);
        in.close();
        System.out.println("\n========================================="
                + "==================================\n");
        /**************************************************************
        **
        **              Input makhluk-makhluk yang ada
        **
        **************************************************************/
        String input = "";
        do {
            try {
                System.out.println("Masukkan sebuah string "
                        + "yang berisi ID (UPPERCASE) makhluk"
                        + "-makhluk yang ingin dimunculkan "
                        + "pada dunia. ");
                System.out.println("Pilihan :\n\tP = PolarBear,"
                        + " \n\tG = Tumbuhan, \n\tR = Rabbit,"
                        + " \n\tT = Turtle, \n\tS = Sheep, "
                        + "\n\tW = Wolf, \n\tU = Snake");
                System.out.print("String input = ");
                input = in.next();
                if (!isValidID(input)) {
                    throw new ExceptionObject(1);
                }
            } catch (ExceptionObject e) {
                e.displayErrorMessage();
            }
        } while (!isValidID(input));
        WorldBuilder.getBuilderInstance().setStrMakhluk(input);
        WorldBuilder.getBuilderInstance().buildWorldObject();
        System.out.println("\n========================================"
                     + "===================================\n");
        /**************************************************************
         **
         **                       Nyalakan dunia
         **
         **************************************************************/
        Thread viewThread = new Thread(new Screen());
        viewThread.start();
        do {
            while (World.getWorldInstance().isPaused() == 1) {
                Thread.sleep(100);
            }
            MakhlukLive.getInstance().makhlukMove();
            MakhlukLive.getInstance().makhlukEat();
            if (World.getWorldInstance().getObjects().isAllMakhlukDead()) {
                World.getWorldInstance().endWorld();
                break;
            }
            Thread.sleep(500);
        } while (World.getWorldInstance().isEnded() == 0);
    }
}
